File: java/chaincue-real-estate/src/main/java/com/chaincuerealestate/chaincuerealestate/routes/house/HousePage.java
package com.chaincuerealestate.chaincuerealestate.routes.house;

import com.chaincuerealestate.chaincuerealestate.domains.Broker;
import com.chaincuerealestate.chaincuerealestate.domains.House;
import com.chaincuerealestate.chaincuerealestate.domains.HouseImage;
import com.chaincuerealestate.chaincuerealestate.routes.home.HomePage;
import com.chaincuerealestate.chaincuerealestate.routes.home.HomePageDTO;
import com.chaincuerealestate.chaincuerealestate.services.DTOBuilderHelpers.CountryHelper;
import com.chaincuerealestate.chaincuerealestate.services.DTOBuilderHelpers.HouseHelper;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Stream;

@RestController
@RequestMapping("house")
@RequiredArgsConstructor
@Slf4j
@CrossOrigin(origins = "*")
public class HousePage {
    private final HouseHelper houseHelper;

    @GetMapping("{houseId}")
    public ResponseEntity<HousePageDTO> housePage(@PathVariable String houseId) {
        log.info("housePage");
        var toDTO = toHomePageDTO(houseId, Optional.empty());
        return ResponseEntity.ok(toDTO);
    }

    private HousePageDTO toHomePageDTO(String houseId, Optional<Function<DTOBuilder, DTOBuilder>> additionalProcessing) {
        return Stream.of(new DTOBuilder())
                .map(additionalProcessing.orElseGet(Function::identity))
                .map(houseHelper.updateDTOBuilderWithHouseByHouseId(houseId, DTOBuilder::setHouse))
                .map(HousePage::toDTO)
                .findFirst().get();
    }

    private static HousePageDTO toDTO(DTOBuilder dtoBuilder) {
        return new HousePageDTO(
                dtoBuilder.getHouse().getId(),
                dtoBuilder.getHouse().getTitle(),
                dtoBuilder.getHouse().getHouseTypes(),
                dtoBuilder.getHouse().getLocation(),
                dtoBuilder.getHouse().getNumberRooms(),
                dtoBuilder.getHouse().getBeds(),
                dtoBuilder.getHouse().getPrice(),
                "₿32.346",
                dtoBuilder.getHouse().getDescription(),
                dtoBuilder.getHouse().getImages().stream().map(HousePage::toDTO).toArray(HousePageDTO.HouseImage[]::new),
                Optional.ofNullable(dtoBuilder.getHouse().getBroker()).map(HousePage::toDTO).orElse(null)
        );
    }

    private static HousePageDTO.HouseImage toDTO(HouseImage houseImage) {
        return new HousePageDTO.HouseImage(
                houseImage.getId(),
                houseImage.getUrl()
        );
    }

    private static HousePageDTO.Broker toDTO(Broker broker) {
        return new HousePageDTO.Broker(
                broker.getId(),
                broker.getName(),
                broker.getPhoneNumber(),
                broker.getEmail()
        );
    }

    @Data
    private static class DTOBuilder {
        private House house;

        public DTOBuilder() {
        }

    }
}

File: kotlin/chaincue-real-estate/src/main/kotlin/com/example/chaincuerealestate/routes/house/HousePage.kt
package com.example.chaincuerealestate.routes.house

import com.example.chaincuerealestate.domains.Broker
import com.example.chaincuerealestate.domains.House
import com.example.chaincuerealestate.domains.HouseImage
import com.example.chaincuerealestate.routes.home.HomePage
import com.example.chaincuerealestate.services.DTOBuilderHelpers.HouseHelper
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("house")
class HousePage (
    private val houseHelper: HouseHelper,
) {

    @GetMapping("{houseId}")
    fun homePage(@PathVariable houseId: String): ResponseEntity<HousePageDTO> {
        log.info("HomePage")
        val toDTO = toHomePageDTO(houseId) { it }
        return ResponseEntity.ok(toDTO)
    }

    private fun toHomePageDTO(houseId: String, additionalProcessing: ((DTOBuilder) -> DTOBuilder)?): HousePageDTO {
        return (additionalProcessing?.invoke(DTOBuilder(null)) ?: DTOBuilder(null))
            .apply { houseHelper.updateDTOBuilderWithHouseByHouseId(houseId) { dtoBuilder: DTOBuilder, house -> dtoBuilder.house = house }.invoke(this) }
            .let { toDTO(it) }
    }

    private fun toDTO(dtoBuilder: DTOBuilder): HousePageDTO {
        return HousePageDTO(
            id = dtoBuilder.house!!.id,
            title = dtoBuilder.house!!.title,
            type = dtoBuilder.house!!.houseTypes,
            location = dtoBuilder.house!!.location,
            numberOfRooms = dtoBuilder.house!!.numberRooms,
            beds = dtoBuilder.house!!.beds,
            dollarPrice = dtoBuilder.house!!.price,
            cryptoPrice = "₿32.346",
            description = dtoBuilder.house!!.description,
            images = dtoBuilder.house!!.images.map { toDTO(it) }.toTypedArray(),
            broker = dtoBuilder.house!!.broker?.let { toDTO(it) }
        )
    }

    private fun toDTO(country: HouseImage): HousePageDTO.HouseImage {
        return HousePageDTO.HouseImage(
            country.id,
            country.url,
        )
    }

    private fun toDTO(broker: Broker): HousePageDTO.Broker {
        return HousePageDTO.Broker(
            broker.id,
            broker.name,
            broker.phoneNumber,
            broker.email
        )
    }

    private data class DTOBuilder(
        var house: House?
    )

    private companion object {
        private val log = LoggerFactory.getLogger(HomePage::class.java)
    }

}
